src/once_cell.rs

```diff
@@ -0,0 +1,800 @@
+use std::cell::UnsafeCell;
+use std::convert::Infallible;
+use std::fmt;
+use std::future::Future;
+use std::mem::{forget, MaybeUninit};
+use std::pin::Pin;
+use std::ptr;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::task::{Context, Poll, RawWaker, RawWakerVTable, Waker};
+
+use event_listener::{Event, EventListener};
+use futures_lite::future;
+
+/// The `OnceCell` is uninitialized.
+const UNINITIALIZED: usize = 0;
+/// The `OnceCell` is being initialized.
+const INITIALIZING: usize = 1;
+/// The `OnceCell` is initialized.
+const INITIALIZED: usize = 2;
+
+/// A memory location that can be written to at most once.
+///
+/// A `OnceCell` can be used to store a single value, and only once. However,
+/// once the value is stored, it can be accessed directly through a reference
+/// instead of needing an RAII guard like `Mutex` or `RwLock`.
+///
+/// # Examples
+///
+/// This structure is useful for a variety of patterns, most notably for one-time
+/// initialization.
+///
+/// ```rust
+/// use async_lock::OnceCell;
+///
+/// # struct Foobar;
+///
+/// async fn very_expensive_initialization() -> Foobar {
+///     // Imagine this is very expensive to initialize,
+///     // for instance, it requires a network request or
+///     // a database call.
+///     # Foobar
+/// }
+///
+/// struct LazyFoobar {
+///     inner: OnceCell<Foobar>,
+/// }
+///
+/// impl LazyFoobar {
+///     fn new() -> Self {
+///         Self {
+///             inner: OnceCell::new(),
+///         }
+///     }
+///
+///     async fn load(&self) -> &Foobar {
+///         self.inner.get_or_init(|| async {
+///             very_expensive_initialization().await
+///         }).await
+///     }
+/// }
+/// ```
+pub struct OnceCell<T> {
+    /// Listeners waiting for a chance to initialize the cell.
+    ///
+    /// These are the users of get_or_init() and similar functions.
+    active_initializers: Event,
+    /// Listeners waiting for the cell to be initialized.
+    ///
+    /// These are the users of wait().
+    passive_waiters: Event,
+    /// State associated with the cell.
+    state: AtomicUsize,
+    /// The value of the cell.
+    value: UnsafeCell<MaybeUninit<T>>,
```

Comment:
I think you need to add `Drop` impl to drop the value because `MaybeUninit<T>` prevents `T` from being dropped,